### PR TITLE
Feature/reduce build warnings

### PR DESCRIPTION
**Pull request by binarymash**
Fixes / New Feature #

#260

## Proposed Changes

Fixes all build warnings except for...

1. The warnings relating to package analysis. These are because of limitations in the current dotnet core package production infrastructure - see https://github.com/NuGet/Home/issues/4587. Not sure what the status is for the fix of this. We could disable package analysis, but that might hide other problems, so suggest we leave it as is for now.

2. CS0618 (using obsolete code) 

I've also enabled errors for some of the rules, which I forgot to do as part of #253. The rules I've chosen are a subset of the rules I use on my own projects. When I turned on errors for these rules it did not result in any errors in the code, so I'm taking that as proof that these rules are pretty unobtrusive and won't cause much hassle for devs unless they are really following bad practices.

### CODE REVIEW

File: src/Ocelot/Configuration/Repository/FileConfigurationRepository.cs [hunk 1]
```diff
@@ -18,7 +18,7 @@ public FileConfigurationRepository(IHostingEnvironment hostingEnvironment)
             _configFilePath = $"{AppContext.BaseDirectory}/configuration{(string.IsNullOrEmpty(hostingEnvironment.EnvironmentName) ? string.Empty : ".")}{hostingEnvironment.EnvironmentName}.json";
         }
 
-        public async Task<Response<FileConfiguration>> Get()
+        public Task<Response<FileConfiguration>> Get()
```
Comment (TomPallister) [inline src/Ocelot/Configuration/Repository/FileConfigurationRepository.cs]: This method we should probably make 

```csharp
jsonConfiguration = System.IO.File.ReadAllText(_configFilePath);
```

into
```csharp
jsonConfiguration = System.IO.File.ReadAllTextAsync(_configFilePath);
```

Then we dont need the Task.FromResult

File: src/Ocelot/Configuration/Repository/FileConfigurationRepository.cs [hunk 2]
```diff
@@ -45,8 +45,8 @@ public FileConfigurationRepository(IHostingEnvironment hostingEnvironment)
 
                 System.IO.File.WriteAllText(_configFilePath, jsonConfiguration);
```
Comment (TomPallister) [inline src/Ocelot/Configuration/Repository/FileConfigurationRepository.cs]: Same as above make this WriteAllTextAsync and we dont need Task.FromResult

**Comment by TomPallister** [comment]
@binarymash sorry took me ages to look at this, been super busy with the new job!

**Comment by binarymash** [comment]
No worries, I'll take a look at this. Hope the new job is going ok, and you're not missing the Pig too much ;)

**Comment by binarymash** [comment]
Ok, this is a bit trickier than the 2 minute job I thought it would be because you can't use await within a lock :)

I'm assuming there's a lock there to prevent concurrent access to the config file, so I looked at locking the file and accessing the filestream via File.Open(...). But, in doing this I see random test failures around accessing the config files. Pretty sure I've seen these failures before, so not sure if it's anything to do with my changes. However, this highlights that changing to use this approach rather than using the sync lock changes the behaviour slightly - we'll now get an exception on concurrent access. Do we ever expect this to happen at runtime? If so, what should we do?

And, what assumptions can we make about the lifetime of the config file? Will it always already exist before we try to write to it? I don't know the operation of Ocelot well enough to understand the exact behaviour here.

Might be worth splitting this into its own issue. What do you reckon?

**Comment by TomPallister** [comment]
@binarymash sorry that is about the 20th time I’ve forgotten about that lock. Feel free to merge!

**Pull request merged**